src/_pytest/main.py

```diff
@@ -470,7 +472,55 @@ def __missing__(self, path: Path) -> str:
 
 
 @final
-class Session(nodes.FSCollector):
+class Dir(nodes.Directory):
+    """Collector of files in a file system directory.
+
+    .. versionadded:: 8.0
+    """
+
+    @classmethod
+    def from_parent(  # type: ignore[override]
+        cls,
+        parent: nodes.Collector,  # type: ignore[override]
+        *,
+        path: Path,
+    ) -> "Dir":
+        """The public constructor.
+
+        :param parent: The parent collector of this Dir.
+        :param path: The directory's path.
+        """
+        return super().from_parent(parent=parent, path=path)  # type: ignore[no-any-return]
+
+    def collect(self) -> Iterable[Union[nodes.Item, nodes.Collector]]:
+        config = self.config
+        col: Optional[nodes.Collector]
+        cols: Sequence[nodes.Collector]
+        for direntry in scandir(self.path):
+            if direntry.is_dir():
+                if direntry.name == "__pycache__":
+                    continue
+                ihook = self.ihook
```

Comment:
`ihook` returns a `FSHookProxy` with a static `remove_mods`, and is thus very sensitive to its placement. It's something I'm hoping to improve in the future (though it's not easy to make something fast for this in pluggy), but with how pytest is now `ihook` cannot be safely hoisted.

Reply:
i believe within the loop it is safe as the remove_mods depends on self.fspath

Reply:
IIRC the exact problem was with conftest plugins that are loaded after the `ihook` and are then not filtered out by `remove_mods`

Reply:
my rough understanding is, with the usage if directories, conftests don't get added until after the collect phase,
as such it presents as safe from my perspective

do we consider conftests in packages at their construction or at their collection?

Reply:
After taking a second and third look at this, I've made a change which removes an ugly hack from this PR, allows the hoisting, and **removes conftest loading from `gethookproxy`** << this has been a real pain in my neck.

Recap (a little complicated, sorry):

### Current main

In current main, calls to `gethookproxy`/`ihook` are the trigger for loading non-initial conftests. This basically means that conftest loading is done almost as a random side-effect, uncontrolled and very non-obvious. And it also dashes any hope of making `gethookproxy` faster (`gethookproxy` shows up prominently in pytest profiles).

I've wanted to improve this for a while, #11268 was the latest step towards that.

### PR before change

In this PR, I ran into a problem.

Previously, `Session` and `Package` did all of the directory traversals inside of their `collect`, which loaded the conftests as a side effect. If the conftest loading failed, it will occur inside of the `collect()` and cause it to be reported as a failure.

Now I've changed things such that `Session.collect` and `Package.collect` no longer recurse, but just collect their immediate descendants, and `genitems` does the recursive expansion work.

The problem though is that `genitems()` doesn't run inside of specific collector's `collect` context. So when it loads a conftest, and the conftest loading fails, the exception isn't handled by any `CollectReport` and causes an internal error instead.

The way I've fixed this problem is by loading the conftests eagerly in a `pytest_collect_directory` post-wrapper, but only during `genitems` to make sure the directory is actually selected.

This solution in turn caused the conftests to be collected *too early*; specifically, the plugins are loaded during the parent's `collect()`, one after the other as the directory entries are collected. So when the `ihook` is hoisted out of the loop, new plugins are loaded inside the loop, and due to the way the hook proxy works, they are added to the `ihook` even though they're supposed to be scoped to the child collectors. So no hoisting.

### PR after change

Now I've come up with a better solution: since now the collection tree actually reflects the filesystem tree, what we really want is to load the conftest of a directory right before we run its `collect()`. A conftest can affect a directory's `collect()` (e.g. with a `pytest_ignore_collect` hookimpl), but it cannot affect how the directory node itself is collected. So I just moved the conftest loading to be done right before calling `collect`, but still inside the `CollectReport` context.

This allows the hoisting, and also removes conftest loading from `gethookproxy` since it's no longer necessary. And it will probably enable further cleanups. So I'm happy with it.